src/main/java/com/teragrep/rlp_03/context/frame/access/Access.java

```diff
@@ -1,79 +1,45 @@
 package com.teragrep.rlp_03.context.frame.access;
 
-import java.util.concurrent.locks.Lock;
-import java.util.concurrent.locks.ReentrantLock;
+import java.util.concurrent.Phaser;
 import java.util.function.Supplier;
 
 public final class Access implements Supplier<Lease> {
-
-    private long accessCount;
-    private boolean terminated;
-    private final Lock lock;
+    final Phaser phaser;
     public Access() {
-        this.accessCount = 0; // TODO consider using a semaphore
-        this.terminated = false;
-        this.lock = new ReentrantLock();
+        // initial registered parties is 1, same as calling phaser.register()
+        this.phaser = new Phaser(1);
     }
 
     @Override
     public Lease get() {
-        lock.lock();
-        try {
-            if (terminated()) {
-                throw new IllegalStateException("Access already terminated");
-            }
-
-            accessCount++;
-            return new Lease(this);
-        } finally {
-            lock.unlock();
+        if (phaser.isTerminated()) {
+            // phaser was already closed by releasing all leases
+            throw new IllegalStateException("Access phaser already terminated");
         }
-    }
 
-    public void terminate() {
-        if (lock.tryLock()) {
-            try {
-                if (accessCount != 0) {
-                    throw new IllegalStateException("Open leases still exist");
-                } else {
-                    if (terminated) {
-                        throw new IllegalStateException("Access already terminated");
-                    }
-                    terminated = true;
-                }
-            } finally {
-                lock.unlock();
-            }
-        }
-        else {
-            throw new IllegalStateException("Lease operation in progress");
-        }
+        phaser.register(); // register new lease to phaser
+        return new Lease(this);
     }
 
-    public boolean terminated() {
-        lock.lock();
-        try {
-            return terminated;
-        }
-        finally {
-            lock.unlock();
+    public void terminate() throws IllegalStateException {
+        phaser.arriveAndDeregister(); // registered=0, should terminate phaser
```

Comment:
i think these should rather check the return value of phaser.arriveAndDeregister(); being non-negative, because terminated phaser returns negative ones